File: src/com/programmers/algorithm/P_43163.java
package com.programmers.algorithm;

public class P_43163 {
    static int initialValue = 51;
    static int min = initialValue;
    public static void main(String[] args) {
        String begin = "hit";
        String target = "cog";
        String[] words =
        //{"hot", "dot", "dog", "lot", "log"};
        {"hot", "dot", "dog", "lot", "log", "cog"};
        solution(begin, target, words);
    }

    public static int solution(String begin, String target, String[] words) {
        dfs(begin, target, words, 0);
        int answer = min == initialValue ? 0 : min;
        System.out.println(answer);
        return answer;
    }

    static void dfs(String begin, String target, String[] words, int depth) {
        if (begin == target) {
            min = depth < min ? depth : min;
            return;
        }

        for (int i = 0; i < words.length; i++) {
            int diffCount = 0;
            String word = words[i];

            for (int j = 0; j < word.length(); j++) {
                if (begin.charAt(j) != word.charAt(j)) {
                    diffCount++;
                }
            }

            if (diffCount == 1) {
                words[i] = "";
                dfs(word, target, words, depth+1);
                words[i] = word;
            }
        }
    }
}

File: Swfit/Algorithm.playground/Sources/Programmers/P_43163.swift
import Foundation

public class P_43163 {
    var initialValue: Int = 51
    var min: Int = 51
    
    public init() {
        
    }
    
    /**
     let begin = "hit"
     let target = "cog"
     var words = ["hot", "dot", "dog", "lot", "log", "cog"]
     */
    
    public func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
        dfs(begin, target, words, 0)
        let answer = min == initialValue ? 0 : min;
        return answer
    }

    func dfs(_ begin: String, _ target: String, _ words: [String], _ depth: Int) {
        var words = words

        if (begin == target) {
            min = depth < min ? depth : min;
            return;
        }
        
        for i in (0..<words.count) {
            let word = words[i]
            var diffCount = 0
            
            let wordCharacters = Array(word)
            let beginCharacters = Array(begin)
            
       
            for j in (0..<wordCharacters.count) {
                if (wordCharacters[j] != beginCharacters[j]) {
                    diffCount += 1
                }
            }
            if (diffCount == 1) {
                words[i] = "";
                dfs(word, target, words, depth+1);
            }
        }
    }

}
